Add cache-aligned atomic

bartoszmodelski opened the pull request:
Cache-aligned atomic to avoid false-sharing in lock-free algorithms. This PR adds another constructor for `'a Atomic.t`, which allocates a 7-word block directly in the shared heap, which is memory aligned by design.

## The design
My main goal here is to get cache-aligned allocations in a way that fits best with the rest of runtime. On the issue, @stedolan suggested to rely on the fact that pools in the slab allocator are already cache line aligned (https://github.com/ocaml/ocaml/issues/11986#issuecomment-1412486056). Since there's a sizeclass for 8-word (1+7) allocations only, each of the allocations there spans exactly cache line, and thus all allocations there are aligned by design. 
```
n                 n+8w               n+16w  
| variable 1      | variable 2       | ... 
```

Some experimenting has confirmed the general idea above, i.e. pools are aligned and consecutive 8-word allocations are spaced out as desired. However, they are not aligned. Pool begins with a 4-word header and that throws the allocations off. We end up with the following:
```
n                 n+8w               n+16w  
| header | variable 1       | variable 2         
```
Therefore, the natural thing to do here seem to double the size of the header. That lands us with the right-looking outcome as far as my tests on x86-64. 

### Update May 12

As pointed out by @gadmm, rather than inflate the header size, we can move the trailing wastage area to the front and things work out fine. Structure of the entire pool looks as follows:
```
n                 n+8w               n+16w     ...       n+3996w    n+4kw
| header   | variable 1      | variable 2 |    ...       |  waste   |     
```

Inflating the header to shift cells is correct but the header has to be 256b (largest cache line sz) for all sizeclasses. Thus, every pool contains ~0.8% wasted space to begin with. We can do the following instead:

```
n                 n+8w               n+16w        n+24w    ...      n+4kw
| header | waste  | variable 1       | variable 2 |        ...      |     
```

With this approach the blocks of 8w, 16w, 32w size are now aligned to the end of the pool and hence aligned to cache lines. We also keep the wastage to the minimum allowed by given sizeclass. 

## Compaction 

In this PR I assume that compaction should not impact this. Intuitively, even on compaction an allocation should be moved to the slab with the same sizeclass, what gives us the same properties. Happy to try to dig more. 

## Alignment

As @smuenzel pointed out, 64b alignment is not enough in the general case. 
* M1's and others have 128b cache lines.
* z/Arch processors have up to 256b (e.g. [IBM z15](https://ibm-zcouncil.com/wp-content/uploads/2020/01/IBM-z15-Technical-Pitch-Handout.pdf)). 
* Apparently, 128b might be beneficial on 64bit in some cases as well ([stackoverflow](https://stackoverflow.com/questions/29199779/false-sharing-and-128-byte-alignment-padding)). 

Based on this, I think that we want at least 128b alignment as minimum. Do we want to push it to 256b? 

Two details here: 
* We could increase the header to 256b and do the actual allocations based on detected archicture/cache line. I cannot speak for how hard it is going to be to get right for all the cases.
* Currently, there is no exclusive sizeclass for 256b. Again, the strongest claim I'm going to make is that adding it manually makes my tests pass.





Related issue https://github.com/ocaml/ocaml/issues/11986.

cc @stedolan @sadiqj @polytypic

bartoszmodelski committed:
Add cache-aligned atomic

Cache-aligned atomic to avoid false-sharing in lock-free algorithms.
This patch relies on the slab allocator in shared heap, which is
cache-line-aligned by design.

gasche commented on runtime/alloc.c:
```diff
@@ -371,3 +371,25 @@ CAMLexport value caml_alloc_some(value v)
+  value var = caml_alloc_shr(7, 0);
+  CAMLassert((var - 8) % Allocation_alignment == 0);
+  caml_initialize(&Field(var, 0), val);
+  return var;
```
I would expect that we have to initialize/zero-out the padding fields, as is done in `caml_alloc`.

bartoszmodelski replied:
Added 0-init for remaining fields

xavierleroy replied:
Zero is not a valid value.

kayceesrk replied:
```
    caml_initialize(&Field(var, i), 0);
```

should be

```
    caml_initialize(&Field(var, i), Val_unit);
```

bartoszmodelski replied:
Thanks for pointing this out. I went with the implementation suggested by @xavierleroy below - marking this as resolved.

gasche commented on runtime/caml/alloc.h:
```diff
@@ -86,6 +87,10 @@ Caml_inline value caml_field_boxed (value arg) { return Field (arg, 0); }
 
+/* Alloc aligned for atomics */
+CAMLextern value caml_alloc_aligned(value);
+CAMLextern value caml_is_aligned(value);
```
I'm not fond of the naming here, "aligned" and "is_aligned" are very generic names that I would expect to be parametrized over an arbitrary alignment (as `caml_aligned_malloc` does for example). Also, the naming just suggests that we are allocating an OCaml value (so I would expect the parameters to be a header and a size or something), but in fact we are specifically allocating a one-world block of tag 0.

Maybe `caml_make_atomic_aligned` or `caml_make_aligned_atomic` would be better names? (And then `caml_atomic_is_aligned` or something. But I agree that you could skip this part of the PR that is not as clearly motivated.)

gasche commented on runtime/caml/config.h:
```diff
@@ -266,4 +266,7 @@ typedef uint64_t uintnat;
 #define Default_runtime_events_log_wsize 16
 
+/* Memory-alignment for allocating atomics */
+#define Allocation_alignment 64
```
Why 64? I guess that we are trying to use the cache-line size to avoid false sharing. This could be documented.

smuenzel replied:
Some quick googling suggests that the Apple M1 and POWER architecture processors have 128b cache line size, and z/architecture processors use 256b.

kayceesrk replied:
I would also suggest adding documentation to explain what "64" is supposed to mean here. Also, consider renaming this to `Allocation_alignment_bsize` (if it indeed is representing size in bytes).

bartoszmodelski replied:
Thank you for suggestions.
* Renamed it.
* Added better documentation.
* Made it default to 128b and added a 256b case for z/Arch.

gasche commented on runtime/shared_heap.c:
```diff
@@ -47,6 +47,7 @@ typedef struct pool {
   value* next_obj;
   caml_domain_state* owner;
   sizeclass sz;
+  long p1; long p2; long p3; long p4; /* padding to align slabs */
```
I'm no C expert, but my impression is that this code is making assumptions about the word size of `long` that may not hold under all support configurations (at least if we include bytecode-only targets). I suspect that there is a caml-runtime-specific type alias that is more appropriate, maybe just `uintnat`. But should we also redefine `sizeclass` to be `uintnat` if we want to be able to make such padding assumptions?

Also, should this be `uintnat padding[4]` instead?

Finally, does this padding-to-align logic also work on systems with 32bit machine words? I think that you are asserting that the first element in a pool is 64-bytes-aligned, and assuming that a 8-word headers ensures that, which may not hold on 32bit systems.

gasche commented on testsuite/tests/callback/callback_effects_gc.ml:
```diff
@@ -1,6 +1,5 @@
 (* TEST
  ocamlrunparam += ",s=512";
- native;
```
I'm not sure why this is part of the PR.

bartoszmodelski replied:
Yes, sorry for the noise

gasche reviewed (commented):
I had a quick look at this PR. I am far from an expert on this sort of bit-level C programming, so some of my questions/remarks may be naive, apologies. My overall feeling (as a non-expert) is that the proposed interface (a new creation function for `Atomic.t`) is very reasonable -- I would even say that I like its simplicity -- but that its implementation in the runtime is not ready.

smuenzel commented on runtime/alloc.c:
```diff
@@ -371,3 +371,25 @@ CAMLexport value caml_alloc_some(value v)
+
+CAMLprim value caml_alloc_aligned(value val)
+{
+  value var = caml_alloc_shr(7, 0);
+  CAMLassert((var - 8) % Allocation_alignment == 0);
```
Shouldn't this depend on Allocation_alignment?

```suggestion
  mlsize_t wo_total_block_size = Allocation_alignment/sizeof(value);
  value var = caml_alloc_shr(wo_total_block_size-1, 0);
  CAMLassert((var - wo_total_block_size) % Allocation_alignment == 0);
```

xavierleroy replied:
This makes even less sense than the original.  I think `var -8` in the original code was trying to get the address of the block header.  Here, `var - wo_total_block_size` subtracts a number of words from the integer representation of a pointer, and it doesn't mean anything to me.

smuenzel replied:
Ah yes, it should have been `sizeof(value)`

xavierleroy replied:
In other words: 8 means at least two different things in the original code.

smuenzel commented on runtime/alloc.c:
```diff
@@ -371,3 +371,25 @@ CAMLexport value caml_alloc_some(value v)
+
+CAMLprim value caml_is_aligned(value val)
+{
+  if ((val - 8) % Allocation_alignment == 0) {
```
same as above, `8` shouldn't be a function of `Allocation_alignment`

smuenzel commented on runtime/caml/sizeclasses.h:
```diff
@@ -1,6 +1,6 @@
 /* This file is generated by tools/gen_sizeclasses.ml */
 #define POOL_WSIZE 4096
-#define POOL_HEADER_WSIZE 4
+#define POOL_HEADER_WSIZE 8
```
This is a generated file, I think you need to change `tools/gen_sizeclasses.ml`

smuenzel replied:
I believe this might be an issue that's breaking the tests, because other values depend on `POOL_HEADER_WSIZE`

xavierleroy commented:
Before we shred the proposed code to pieces, could we first agree on a design?  I think the proposed API on the OCaml side makes sense.  I don't understand how it interacts with the current shared-heap allocator.

xavierleroy commented:
I marked this PR as "Draft", because I don't think it is ready for review.  Let's articulate a design first, before reviewing an implementation.

bartoszmodelski commented:
Thank you all for the feedback! 

I've moved the design from here to PR description.

smuenzel commented:
I think that some of your explanations say "byte", when you mean "word". Can you edit/update the design document?

bartoszmodelski commented:
Thanks, corrected.

kayceesrk commented:
> which allocates a 7-byte block

Should this be `7-word`?

bartoszmodelski commented:
Thank you, corrected.

kayceesrk commented on runtime/alloc.c:
```diff
@@ -371,3 +371,27 @@ CAMLexport value caml_alloc_some(value v)
+
+CAMLprim value caml_atomic_is_aligned(value val)
+{
+  if ((val - 8) % Allocation_alignment == 0) {
```
The `8` here needs to be fixed in the same way as `caml_make_atomic_aligned`. `Hp_val` macro may make it explicit what you are trying to do here:

https://github.com/ocaml/ocaml/blob/f11f82f18993f969be5cf5bc4a0c133a053949cb/runtime/caml/mlvalues.h#L162

It returns the pointer to the header given the value on the heap.

bartoszmodelski replied:
Change above code code to use `Hp_val` (the function is now defined in the test).

kayceesrk commented on runtime/alloc.c:
```diff
@@ -371,3 +371,27 @@ CAMLexport value caml_alloc_some(value v)
+
+/* For aligned atomic */
+
+CAML_STATIC_ASSERT((sizeof(value)) <= 8);
```
Where does the "8" come from? Why is this important here? 

Generally, it would be best to avoid hard-coded constants in the code.

bartoszmodelski replied:
Removed this assertion altogether. It doesn't seem needed, in hindsight.

kayceesrk commented on runtime/shared_heap.c:
```diff
@@ -47,6 +47,7 @@ typedef struct pool {
   sizeclass sz;
+  uintnat p[4];  /* padding to align the beginning of slabs to cache line */
 } pool;
 CAML_STATIC_ASSERT(sizeof(pool) == Bsize_wsize(POOL_HEADER_WSIZE));
```
Can you also consider adding a static assert to relate the size of pool to the allocation alignment.

```
CAML_STATIC_ASSERT(sizeof(pool) == Wsize_bsize(Allocation_alignment_bsize));
```

bartoszmodelski replied:
Is `Wsize_bsize` needed? We have bytes on both sides. I think LHS will be halved on 32-bit, so we could do the same for RHS. For now, I've added the following, which is a bit weaker: 

`CAML_STATIC_ASSERT(sizeof(pool) % Cache_line_bsize == 0);`

kayceesrk replied:
Ok. You are right about bytes on both sides.

The new assertion works for me.

kayceesrk commented on stdlib/atomic.mli:
```diff
@@ -55,6 +58,10 @@ val incr : int t -> unit
 
+(** [is_aligned r] checks cache-alignment of atomic variable. For references
+    allocated with [make], the output may change mid-execution. *)
+val is_aligned : 'a t -> bool
```
Why is this needed? 

If this is only needed for the testcase, may I suggest implementing the C function `caml_atomic_is_aligned` in the testcase. There are a number of examples in the testsuite which shows how to implement test cases which involve both OCaml and C code. For example, see `testsuite/tests/callback/test2.ml`.

bartoszmodelski replied:
Moved `is_aligned` to tests. Thanks for the example!

xavierleroy commented:
Thanks for the design notes.

> Some experimenting has confirmed the general idea above, i.e. pools are aligned and consecutive 8-word allocations are spaced out as desired. However, they are not aligned. Pool begins with a 4-word header and that throws the allocations off. 

Is this really a problem?   EDIT: yes, it is; we get first fields of blocks (those containing the payloads of Atomic.t values) at addresses = 5 mod 8 (words), but addresses = 0 to 3 mod 8 (words) can contain data from a mutable block (but not an Atomic.t value!), which is enough to cause false sharing.

xavierleroy commented:
At any rate,  here is a simplified implementation.  The remaining issue  is to define `Cache_line_bsize` (size of cache lines in bytes) appropriately for the target processor.
```
#define Cache_line_bsize 64  /* or whatever is appropriate for the platform */

CAMLprim value caml_atomic_make_aligned (value v)
{
  CAMLparam1(v);
  const mlsize_t sz = Wosize_bhsize(Cache_line_bsize);
  value res = caml_alloc_shr(sz, 0);
  caml_initialize(&Field(res, 0), v);
  for (mlsize_t i = 1; i < sz; i++) Field(res, i) = Val_unit;
  CAMLreturn(res);
}
```

xavierleroy commented:
Re: alignment of allocations in major heap pools, instead of adding fixed padding to the `pool` structure, you could perhaps change `pool_initialize`
https://github.com/ocaml/ocaml/blob/23dab79a4e42856aa33816b9c79c3d4d79959cb9/runtime/shared_heap.c#L233-L239
so that the initial `p` is `sz`-aligned, at least if `sz` s a power of 2.  Something like this:
```
  mlsize_t ofs0 = sz > POOL_HEADER_SZ ? sz : POOL_HEADER_SZ;
  value* p = (value*)((char*)r + ofs0);
```
or, if you like Hacker's Delight and want to stick to the spec, nothing but the spec:
```
  mlsize_t ofs0 = sz > POOL_HEADER_SZ && sz & (sz - 1) == 0 ? sz : POOL_HEADER_SZ;
  value* p = (value*)((char*)r + ofs0);
```

sadiqj commented:
One thing to note, the proposed implementation of the compactor in https://github.com/ocaml/ocaml/pull/12193 adds a flag to the shared pool header: https://github.com/ocaml/ocaml/pull/12193/files#diff-b468e47c34ab18ea7b2cd8322cde20a2fa4d3e647422ee725f0fe7d84f768a32R54 which might make the decision of how to do alignment easier.

bartoszmodelski commented:
I've pushed a number of changes: 
* Use @xavierleroy's implementation for atomic_make_aligned.
* Move is_aligned to testsuite.
* Use 128 bytes as the default alignment, and 256 bytes for z/Arch. 
    * Increase header size to 256 bytes.
    * Add sizeclass of size 32 words. Tweaked max_overhead slightly, to keep the number of sizeclasses at 32. (Please let me know if I should re-run any benchmarks). 
* Better names, comments, etc.


One last thing I'm not sure about, is how to do fix the header size. The suggestion to do the change in `pool_initialize` to shift slabs to alignment looks good but I think that `gen_sizeclasses.ml` needs to be aware of this to properly compute wastage (and possibly other things).

Perhaps I could introduce a POOL_HEADER_MAX_WSIZE constant (or POOL_HEADER_RESERVED_WSIZE) that's treated as the de facto size of pool header (shifting the allocation of slabs as desired)? 

FWIW, another thing that seems to work is attaching `CAMLalign` on the first field of `pool` struct. But we'd have to wrap it to work with words rather than bytes.

bartoszmodelski commented on runtime/caml/config.h:
```diff
@@ -266,4 +266,15 @@ typedef uint64_t uintnat;
+   /* Many 64-bit processors have 128-byte cache line. Even with 64-byte cache
+   line, it may be beneficial to have twice that. */
+   #define Cache_line_bsize 128
+#endif
```
Perhaps, a better name here is `Assumed_cache_line_bsize`?

kayceesrk replied:
The comment clearly says that this is the assumed size of the cache line. So the name is fine.

kayceesrk commented:
FWIW, it would be useful to add your branch to Sandmark nightly https://github.com/ocaml-bench/sandmark-nightly-config#adding-your-compiler-branch-to-the-nightly-runs to see if it affects the performance. CC @shakthimaan.

kayceesrk commented on runtime/alloc.c:
```diff
@@ -371,3 +371,13 @@ CAMLexport value caml_alloc_some(value v)
+{
+  CAMLparam1(v);
+  const mlsize_t sz = Wosize_bhsize(Cache_line_bsize);
+  value res = caml_alloc_shr(sz, 0);
```
Do you not want `sz - 1` here to account for the fact that we have a 1-word header? I remember that the earlier code did `caml_alloc_shr(7,0)`. What am I missing?

kayceesrk replied:
Answering the question myself -- `Wosize_bhsize` and not `Wosize_bsize`! 👍

kayceesrk commented on tools/gen_sizeclasses.ml:
```diff
@@ -15,12 +15,27 @@
+  Prevention of false sharing requires certain sizeclasses to be present. This
+  ensures they are generated.
+
+  Runtime has a constructor for atomics, which aligns them with cache lines
```
```suggestion
  Runtime has a constructor for atomics (`caml_atomic_make_aligned `), which aligns them with cache lines
```

kayceesrk commented on tools/gen_sizeclasses.ml:
```diff
@@ -15,12 +15,27 @@
+
+  Runtime has a constructor for atomics, which aligns them with cache lines
+  to avoid false sharing. The implementation relies on the fact that pools are
+  allocated with high alignment and slots of appropriate size are laid out in
```
What does "high alignment" mean?

bartoszmodelski replied:
improved this comment

kayceesrk commented on tools/gen_sizeclasses.ml:
```diff
@@ -15,12 +15,27 @@
+  Runtime has a constructor for atomics, which aligns them with cache lines
+  to avoid false sharing. The implementation relies on the fact that pools are
+  allocated with high alignment and slots of appropriate size are laid out in
+  such a way, that their borders coincide with borders between cache lines.
```
s/borders/boundaries

kayceesrk commented:
A question on design. I understand that the goal is to avoid false sharing on the atomic values by padding it appropriately. In the current design, the padding has two costs -- memory and GC. The memory cost of padding is unavoidable and is by design. However, the padding also means that the GC will have to trace the other fields, where it will only find `Val_unit`. Is there a way to avoid this at all? (Not easily I presume).

bartoszmodelski commented:
Current patch inflates the atomic block to get a slab of appropriate sizeclass. In theory, we could instead try to explicitly tell the allocator to put this small allocation into a bigger slab. The core part of the patch would look something as follows (+ some code to expose as `caml_alloc_shr_padded`). 

```ocaml
 value* caml_shared_try_alloc(struct caml_heap_state* local, mlsize_t wosize,
-                             tag_t tag, reserved_t reserved, int pinned)
+                             tag_t tag, reserved_t reserved, int pinned,
+                             mlsize_t wpadding)
 {
   mlsize_t whsize = Whsize_wosize(wosize);
   value* p;
   uintnat colour;
 
   CAMLassert (wosize > 0);
+  CAMLassert (wpadding >= 0);
   CAMLassert (tag != Infix_tag);
 
   CAML_EV_ALLOC(wosize);
 
-  if (whsize <= SIZECLASS_MAX) {
+  mlsize_t whalloc_size = whsize + wpadding;
+
+  if (whalloc_size <= SIZECLASS_MAX) {
     struct heap_stats* s;
-    sizeclass sz = sizeclass_wsize[whsize];
-    CAMLassert(wsize_sizeclass[sz] >= whsize);
+    sizeclass sz = sizeclass_wsize[whalloc_size];
+    CAMLassert(wsize_sizeclass[sz] >= whalloc_size);
     p = pool_allocate(local, sz);
     if (!p) return 0;
     s = &local->stats;
     s->pool_live_blocks++;
     s->pool_live_words += whsize;
     s->pool_frag_words += wsize_sizeclass[sz] - whsize;
   } else {
+    CAMLassert(wpadding == 0);
     p = large_allocate(local, Bsize_wsize(whsize));
     if (!p) return 0;
   }
   <initialization of block based on wosize>
```

sadiqj commented:
I'm not sure the cost in tracing is going to be meaningful @kayceesrk . The mark stack representation is a range and https://github.com/ocaml/ocaml/blob/trunk/runtime/major_gc.c#L910 means we'll just end up doing a few more iterations of the loop, nearly all of which will be in cache.

I don't think it's worth adding complexity to avoid it.

kayceesrk commented:
> The mark stack representation is a range and https://github.com/ocaml/ocaml/blob/trunk/runtime/major_gc.c#L910 means we'll just end up doing a few more iterations of the loop, nearly all of which will be in cache.

Ok. Makes sense.

kayceesrk commented:
I am happy with the current design. @xavierleroy what do you think about the current design?

> FWIW, it would be useful to add your branch to Sandmark nightly https://github.com/ocaml-bench/sandmark-nightly-config#adding-your-compiler-branch-to-the-nightly-runs to see if it affects the performance. CC @shakthimaan.

Did you add your branch to sandmark nightly @bartoszmodelski?

gasche commented on tools/gen_sizeclasses.ml:
```diff
@@ -15,12 +15,27 @@
   1. -. float_of_int((block / slot) * obj) /. float_of_int block
 
-let max_overhead = 0.10
+let max_overhead = 0.101
```
why is this value changed?

bartoszmodelski replied:
Adding a 32-word sizeclass forces to script to add 33rd sizeclass at 0.1. This tweak keeps the number of sizeclasses constant.

gasche commented:
@kayceesrk it is not clear to me why you are asking for a benchmark here. The bulk of the change is a new runtime primitive to allocate cache-line-sized `Atomic.t`, which no one is using for now and therefore will not affect performance. There is some tweaking of the "size classes", that is the pools used by the GC for objects of fixed size. Is it the part that we would be benchmarking?

avsm commented:
The changes to the size classes should just go through a Sandmark test to check for any unexpected regressions.  There's plenty of counter intuitive results we've encountered in the past during multicore development from cache behaviours when changing these.

xavierleroy commented on runtime/caml/alloc.h:
```diff
@@ -86,6 +87,10 @@ Caml_inline value caml_field_boxed (value arg) { return Field (arg, 0); }
 #define caml_alloc_unboxable caml_alloc_boxed
 #define caml_field_unboxable caml_field_boxed
 
+/* Special constructor for atomics that should have exclusive ownership of
+  cache line. */
+CAMLextern value caml_atomic_make_aligned(value);
+
```
In general, primitives (functions defined `CAMLprim`) are not declared in header files, because they are called from OCaml code only.

Here, we have a function that could reasonably be called from C code, so it's less clear.  But note for instance that there is no `caml_atomic_make` (unaligned) function that can be called from C, the closest equivalent being `caml_alloc_1(0, v)`.  

Other opinions welcome.

kayceesrk replied:
Good point. We have the following functions to manipulate atomics from C, but not allocation. 

```c
memory.c:CAMLprim value caml_atomic_load (value ref)
memory.c:CAMLprim value caml_atomic_exchange (value ref, value v) /* stores are implemented as exchanges */ 
memory.c:CAMLprim value caml_atomic_cas (value ref, value oldv, value newv)
memory.c:CAMLprim value caml_atomic_fetch_add (value ref, value incr)
```

It may be useful to provide access to OCaml atomics from C, including allocations. This may be done in a separate PR. For this PR, let's remove this declaration from the header file.

bartoszmodelski replied:
This is now removed.

gadmm replied:
In addition I vaguely remember that there is a subtlety memory-model-wise whereby the C version of one of those should be different from the CAMLprim version.

xavierleroy commented on runtime/caml/config.h:
```diff
@@ -266,4 +266,15 @@ typedef uint64_t uintnat;
+#ifdef TARGET_s390x
+   #define Cache_line_bsize 256
+#else
+   /* Many 64-bit processors have 128-byte cache line. Even with 64-byte cache
+   line, it may be beneficial to have twice that. */
```
I bite: why is it beneficial to use 2 cache lines where one would suffice (intuitively)?

bartoszmodelski replied:
L2 spatial prefetcher on Intel 64 prefetches 128-byte chunks. If two domains act on atomics residing in the same chunk and prefetcher kicks in, one will steal others cache line. 
* https://docs.rs/crossbeam/latest/crossbeam/utils/struct.CachePadded.html#size-and-alignment 
* https://stackoverflow.com/questions/29199779/false-sharing-and-128-byte-alignment-padding

Sadly, I don't have any high-quality references. I'm happy to spend a little time trying to prove that it matters for us, and if there's no effect, make it 64-byte on x86 64 (and RISC-V, I think).

gadmm replied:
I wondered the same. The best reference on stackoverflow seems to be https://stackoverflow.com/questions/72126606/should-the-cache-padding-size-of-x86-64-be-128-bytes and reading it in details shows that it is very debated whether the spatial prefetcher influences destructive interference.

For the record I ran the test found there on my machine (AMD Zen 3) and I do not see benefits of choosing 128 over 64. In addition, gcc decided that [std::hardware_constructive_interference_size and std::hardware_destructive_interference_size](https://en.cppreference.com/w/cpp/thread/hardware_destructive_interference_size) should both be 64 on my platform.

If the right value is machine-dependent for the same compilation target, could it be implemented as something else than a compile-time constant?

gadmm replied:
One hypothesis from https://stackoverflow.com/a/72127222 is that without actual false sharing (within the same cache line), there is not effect of the L2 spatial prefetcher (each thread ending up exclusive owner of its cache line). One hypothesis from https://www.mail-archive.com/gcc-patches@gcc.gnu.org/msg265387.html is that the L2 spatial prefetcher only concerns loads from memory to L2 and L3 cache and thus does not cause destructive interference, based on wording from the Intel optimization manual (this is the thread where gcc settled on 64 for the destructive interference size).

I went on and reproduced the tests at https://stackoverflow.com/a/72127222 on Intel i7-1185G7 (Tiger Lake). Starting from the benchmark provided there, I observed:
- No benefit of 128-alignment over 64-alignment if there is at most a thread per cache line. I also did not manage to reproduce the (negligible) effect on the counter `machine_clears.memory_order` mentioned in the answer.
- Moderate benefit of 128-alignment over 64-alignment when: 1) there are two threads competing on each cache line (4 threads total), 2) the operations are RMW (e.g. fetch_add). Moreover, the slowdown is around 20% (compare to 100% slowdown when all 4 threads try to access the same cache line). By running with hardware prefetcher turned off as a control, the slowdown disappears.

As a conclusion, the effect of the spatial prefetcher exists but is hard to reproduce, as it essentially only worsens pre-existing interference. It seems that the appropriate `Cache_line_size` on x86-64 is 64, because in order to avoid destructive interference from cache line pairs, it is enough to avoid destructive interference from single cache lines.

bartoszmodelski replied:
Thank you for doing a deep dive on this. 

I've changed the `config.h` to 64b for amd64, riscv5 and non-native (this case is a little odd, in that I'm not sure if the value is actually being used outside tests).

xavierleroy commented on runtime/shared_heap.c:
```diff
@@ -47,8 +47,10 @@ typedef struct pool {
   value* next_obj;
   caml_domain_state* owner;
   sizeclass sz;
+  uintnat padding[28];
```
For the record: what this padding is doing (and the `#define POOL_HEADER_WSIZE 32`) is to ensure that allocations in the pool start at a 256-aligned address on 64-bit platforms and 128-aligned address on 32-bit platforms.   s390x being a 64-bit platform, this alignment is sufficient, although more than strictly necessary on other platforms.

xavierleroy commented on stdlib/atomic.mli:
```diff
@@ -29,6 +29,10 @@ type !'a t
 (** Create an atomic reference. *)
 val make : 'a -> 'a t
 
+(** Create a cache line-aligned atomic reference. Use it over [make v] to
+    prevent false-sharing. *)
```
Perhaps mention that memory requirements are increased?  Like 8x compared to a normal atomic reference.  If I'm making a list with an atomic reference at each cell, for instance, maybe I don't want the 8x size increase.

bartoszmodelski replied:
Added a mention

xavierleroy reviewed (commented):
This is shaping up nicely.  I'll remove the Draft status to encourage others to review.   Some questions below.

kayceesrk commented:
> There is some tweaking of the "size classes", that is the pools used by the GC for objects of fixed size. Is it the part that we would be benchmarking?

Yes. Only to understand this impact.

bartoszmodelski commented:
@kayceesrk Opened a [PR](https://github.com/ocaml-bench/sandmark-nightly-config/pull/21) against sandmark to add it.

kayceesrk commented:
> Opened a https://github.com/ocaml-bench/sandmark-nightly-config/pull/21 against sandmark to add it.

I've merged it. Results should appear tomorrow at https://sandmark.tarides.com/.

gadmm commented on runtime/caml/sizeclasses.h:
```diff
@@ -1,19 +1,19 @@
 /* This file is generated by tools/gen_sizeclasses.ml */
 #define POOL_WSIZE 4096
-#define POOL_HEADER_WSIZE 4
+#define POOL_HEADER_WSIZE 32
```
An alternative approach that does not involve changing the ~size classes~ _header size_ could be to align the pool contents with the _end_ of the pool (with a padding computed dynamically, or by allocating downwards from the end of the pool). It would also avoid wasting space with padding.

Edit: I just saw your reply to Xavier's similar suggestion of a dynamic padding about updating wastage in sizeclass.h. Seeing it as simply aligning the arena with the end of the pool shows that there is no need to update wastage.

bartoszmodelski replied:
That's a neat idea! I've removed the padding and switched "wastage area" to be between the header and slabs.

gadmm commented on stdlib/atomic.mli:
```diff
@@ -29,6 +29,11 @@ type !'a t
 
+(** Create a cache line-aligned atomic reference. Use it over [make v] to
+    prevent false-sharing. Atomic reference allocated with [make_aligned v]
+    occupies 8-16x the memory of one allocated with [make v]. *)
```
~This would only be 4-8x if you take the header word into account.~

cache line-aligned -> cache-line-aligned

gadmm replied:
Reading the PR again, this is indeed `8-16x`, as 64B alignment is not used currently in the PR. Sorry for the noise.

xavierleroy replied:
> cache line-aligned -> cache-line-aligned

Opinions from native English speakers would be welcome.  A Google search shows occurrences of "cache line aligned", "cache-line aligned", and "cache line-aligned", but not "cache-line-aligned".

(I guess there's a meta-rule that you can use at most one hyphen in a compound adjective.)  

As a non-native speaker, I'd rather chicken out and write "Create an atomic reference that is aligned on a cache line.".

avsm replied:
"Create an atomic reference that is aligned on a cache line" is indeed the clearest of those options.

gadmm replied:
Thanks, much better indeed. Just to clarify, after referring to the Chicago Manual of Style, I stand by "cache line-aligned" being incorrect and "cache-line-aligned" being correct as a form derived from "cache line aligned". Whether the latter phrase sounds entirely right to a native speaker is something else, but it is not a matter of a limit on the number of hyphens one can use. You will find more instances of "cache-line-aligned" on Google scholar.

alainfrisch replied:
> Opinions from native English speakers would be welcome. A Google search shows

This is so 2022...

GPT-4 says:

> The correct usage depends on the style guide that you are following. However, in many English writing contexts, the hyphen is used to link words together to avoid ambiguity. In this particular case, "cache-line-aligned" would be the most appropriate usage because it clearly conveys that the phrase is referring to alignment with respect to cache lines, not an aligned cache or line. The hyphenation creates a single, coherent idea.

> In contrast, "cache line-aligned" could potentially be misread as "cache, line-aligned", which conveys a different meaning. 

> If no specific style guide is being followed, the important thing is to be consistent throughout your text. If you choose to use "cache-line-aligned," make sure you use the same structure throughout your work. 


GPT-3.5 begs to differ:

> The common and widely accepted notation for the term you mentioned is "cache line-aligned," with a space between "cache" and "line." This notation helps clarify that "line-aligned" refers to alignment with the cache line size. It's important to note that style conventions can vary, and different publications or organizations may have their own preferences. However, using "cache line-aligned" is a generally accepted and clear way to express the concept.

bartoszmodelski replied:
Thank you all. I've put Anil's suggestion in place.

gadmm commented on runtime/caml/sizeclasses.h:
```diff
@@ -1,19 +1,19 @@
-{ 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 12, 14, 16, 18, 20, 23, 26, 29, 33, 37, 42,
-  47, 53, 59, 65, 73, 81, 89, 99, 108, 118, 128 };
+{ 1, 2, 3, 4, 5, 6, 7, 8, 10, 12, 14, 16, 17, 19, 22, 25, 28, 32, 34, 38, 43,
+  48, 53, 59, 66, 73, 81, 89, 98, 107, 117, 128 };
```
The use of cache-line-alignment I know involves atomic arrays, used to spread a load among non-interfering atomic values. To have cache-line-aligned atomic arrays, one would need all multiples of 16. Is having cache-line-aligned atomic arrays something you consider necessary eventually?

bartoszmodelski replied:
We have some structures in lockfree, which use arrays of atomics. It would be useful for better caching. But I think it's a bigger lift because atomics are not immediate now (so an array `'a Atomic.t array`, where atomics have been instantiated with `make_aligned` should not experience false-sharing as far as atomic operations are concerned). 

Agreed about having all multiples of 16 there, and perhaps also a special path for large allocations. Either that or a nice wrapper to allocate a bigger array and rely on padding, I suppose.

(this code changed since header size is at 4 byte again)

bartoszmodelski commented:
I've pushed an update: 
* Implemented the suggestion to move wastage block to the front. I've updated the design in PR description to reflect this.  
* Changed `Cache_line_bsize` to 64 for x64, riscv and non-native
* Fixed the langauge in `atomic.mli`

kayceesrk commented:
The benchmarks look fine.

* [Sequential on Navajo](https://sandmark.tarides.com/?app=Sequential+Benchmarks&sequential_12=navajo_5.2.0%2Btrunk%2Btsan%2Bsequential_20230514_63fb28d&sequential_B1=20230514&sequential_B2=%5B%27navajo_5.2.0%2Btrunk%2Btsan%2Bsequential_20230514_63fb28d%27%5D&sequential_10=5.2.0%2Btrunk%2Btsan%2Bsequential&sequential_find_by=variant&sequential_01=20230514&sequential_00=5.2.0%2Btrunk%2Bsequential&sequential_B0=navajo&sequential_02=navajo_5.2.0%2Btrunk%2Bsequential_20230514_70ff202&sequential_num_variants=2&sequential_11=20230514)
  + There is a small variance in both directions for the running time. About 4% in either direction. This I think is within expected range
  + There is a 10-20% increase in maxRSS on the cubicle benchmarks, which is odd. I suspect that this is due to secondary effects of making the pool headers larger. The "top heap words" doesn't vary much between trunk and this PR. So I don't expect to find anything interesting to investigate here. 
* [Parallel on Navajo](https://sandmark.tarides.com/?app=Parallel+Benchmarks&parallel_02=navajo_5.2.0%2Btrunk%2Bparallel_20230516_cd01cac&parallel_num_variants=2&parallel_10=5.2.0%2Btrunk%2Btsan%2Bparallel&parallel_find_by=variant&parallel_12=%5B%27navajo_5.2.0%2Btrunk%2Btsan%2Bparallel_20230516_63fb28d%27%5D&parallel_11=20230516&parallel_00=5.2.0%2Btrunk%2Bparallel&parallel_01=20230516)

bartoszmodelski commented:
>   * There is a 10-20% increase in maxRSS on the cubicle benchmarks, which is odd. I suspect that this is due to secondary effects of making the pool headers larger. The "top heap words" doesn't vary much between trunk and this PR. So I don't expect to find anything interesting to investigate here.

To be precise, the pool header is no longer larger, since wastage is reused as padding. However, with the changes to size of some sizeclasses, in the worst case some allocations may end up in slots 10% larger and that would fit.

gadmm commented:
I find that this PR could benefit from a more detailed explanation of its motivation. Could you provide a compelling, typical use-case scenario for `Atomic.create_aligned`? The example of "making a list with an atomic reference at each cell" leaves me unconvinced that it would be a typical use-case; your experience with lock-free data structures is probably more compelling. Specifically, could you clarify whether you optimally need an atomic that does not interfere with any other memory cell, or an array of atomics that do not interfere with each other? The solution for the latter would be different from a change in size classes. Could you tell us what is provided in other languages for this problem (leaving aside C/C++ and other languages with tight control on memory layout)?

As an outcome it would be nice if we end up with a compelling example in the documentation for `create_aligned`.

bartoszmodelski commented:
Examples: 
* For a single data structure performance Anmol's benchmarks on a simple queue with two shared atomic pointers show a significant improvement coming from enforcing the alignment [Aligned int repo](https://github.com/anmolsahoo25/aligned_int). While the benchmark is simple, the "array with two pointers" design lets us build all kinds of bounded single/multi-producer/consumer stacks/queues/bags. 
* There's also a different issue in that the amount of false-sharing program gets may change in unexpected ways. Thus making the benchmarks far noisier and reactive to seemingly unrelated changes for micro-architectural reasons. https://github.com/ocaml/ocaml/issues/11986#issuecomment-1412418859

I looked at Java and Go a while ago. They both let user add padding in the structure. Java has `@Contended` keyword, while Golang `cpu.CacheLinePad`. 

For C/C++, I think that padding is only required for arrays. For a single variable just using `aligned_alloc` works (which as far as I remember just allocates a larger block and chops it to alignment).

gadmm reviewed (approved):
Thanks for the pointers. Reading about `@Contended` and `cpu.CacheLinePad`, I agree that the present PR offers something like other language's toolbox. And instead of an example, it seems that there is a guideline along the lines of "first measure to see if you observe contention, and then decide if you benefit from alignment".

Regarding the specification, this seems less about being aligned than alone on a cache line (since alignment does not necessarily imply padding): "that is aligned on a cache line" => "that is alone on a cache line".

Apart from these two suggestions to improve the documentation, this PR looks good to me.

bartoszmodelski commented:
Agreed with both.

gasche commented:
I agree with @gadmm's remark on the naming focus on "alone/separate/contended" rather than "aligned". Should it go deeper than the documentation, and suggest a better name than `caml_atomic_make_aligned` and `Atomic.make_aligned`? What about, for example:
- `Atomic.make_isolated`
- `Atomic.make_separated`
- `Atomic.make_cache_alone`
- `Atomic.cache_contended`?

Once this question is resolved (possibly just by telling me that names are hard and we should go with the current version), I would be happy to approve the PR (I have only done a partial review so I would also approve on behalf of @gadmm).

gadmm commented:
`make_aligned` is ok as a colloquial name. `make_padded` is perhaps a bit more accurate.

bartoszmodelski commented:
I've improved the comment. Feel free to push or suggest further changes.

As for the name I find Java-ish _make_contended_ to be the clearest. It references the use-case rather than the implementation details (unlike _make_padded_, _make_aligned_) and uses typical vocab for the area (unlike _make_isolated_, _make_separated_). 
 
Other than that, for _padded_ vs _aligned_, these are the cases I've seen:
* padding means just padding (Go, Java)
* aligned may entail padding ([aligned_alloc](https://en.cppreference.com/w/c/memory/aligned_alloc) requires size to be a multiple of alignment).

gasche commented:
I like `make_contended` slightly better than `make_aligned` and would be in favor of the rename if you also like it. (Hopefully no one dislikes the new name!)

bartoszmodelski commented:
Sg, renamed it.

kayceesrk reviewed (approved):
LGTM

gasche commented on Changes:
```diff
@@ -229,6 +229,9 @@ OCaml 5.1.0
 ### Standard library:
 
+* #: Add cache-aligned atomic.
+  (Bartosz Modelski, review by )
```
(this should be fixed)

gasche commented:
@bartoszmodelski could you fill the Changes entry, so that we could merge this?

A drive-by comment: C++17 has introduced a variable [hardware_destructive_interference_size](https://en.cppreference.com/w/cpp/thread/hardware_destructive_interference_size) to get the required constant in a portable way. If the C compiler we rely on supports this, it may be a better default than our own home-grown logic. (I am not suggesting that we change the current PR to support this, it could be a follow-up and it requires more configure-fu than most people have.)

bartoszmodelski committed:
update Changes

bartoszmodelski commented:
Thanks for the ping. Pushed Changes just now. Feel free to amend the entry if needed.
